cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+	dataDir            string
+	componentReference string
+	lidarConfig        LidarConfig
+}
+
+// CartoAlgoConfig contains config values from app
+type CartoAlgoConfig struct {
+	optimizeOnStart      bool
+	optimizeEveryNNodes  int
+	numRangeData         int
+	missingDataRayLength float64
+	maxRange             float64
+	minRange             float64
+	maxSubmapsToKeep     int
+	freshSubmapsCount    int
+	minCoveredArea       float64
+	minAddedSubmapsCount int
+	occupiedSpaceWeight  float64
+	translationWeight    float64
+	rotationWeight       float64
+}
+
+// NewCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewCartoLib(miniloglevel, verbose int) (CartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(&pVcl))
+	/*
+		Question: what is going on here? are we not getting correct amount of indirection?
+					fmt.Printf("%#v\n", pVcl) = &cartofacade._Ctype_struct_viam_carto_lib{minloglevel:0, verbose:0}
+					fmt.Println(&pVcl) = 0x14000128078
+					fmt.Println(goPtr) = 0x600001ee8000
+					// ppVcl := (**C.viam_carto_lib)(goPtr)
+					fmt.Println(ppVcl) = 0x600001ee8000
+	*/
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return CartoLib{}, err
+	}
+
+	vcl := CartoLib{value: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CartoLib) TerminateCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewCarto(cfg CartoConfig, acfg CartoAlgoConfig, vcl CartoLib) (Carto, error) {
+	pVc := C.alloc_viam_carto()
+	defer C.free_alloc_viam_carto(pVc)
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(&pVc))
+
+	vcc := getConfig(cfg)
+	defer C.free_bstring_array(vcc.sensors)
+
+	vcac := getAlgoConfig(acfg)
+
+	status := C.viam_carto_init(ppVc, vcl.value, vcc, vcac)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return Carto{}, err
+	}
+
+	return Carto{value: (*ppVc)}, nil
+}
+
+// Start is a wrapper for viam_carto_start
+func (vc *Carto) Start(ctx context.Context) error {
+	status := C.viam_carto_start(vc.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Stop is a wrapper for viam_carto_stop
+func (vc *Carto) Stop(ctx context.Context) error {
+	status := C.viam_carto_stop(vc.value)
+
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TerminateCarto calls viam_carto_terminate to clean up memory for viam carto
+func (vc *Carto) TerminateCarto(ctx context.Context) error {
+	status := C.viam_carto_terminate(&vc.value)
```

Comment:
Why does this not require a cast to unsafe pointer? If this works, why are we not using this approach in the lib terminate?

Reply:
updated to use this approach, the unsafe pointer was unnecessary I believe

Reply:
looks great!